### PR TITLE
download batch of scans from s3 first pass

### LINKED ISSUE #10: Download a batch of satellite data from AmazonS3
## Checklist

- Select Project Label
- Select Sub-Project Milestone

## Feature Description

Download a batch of scans from the [noaa-goes16](https://s3.console.aws.amazon.com/s3/buckets/noaa-goes16/?region=us-west-2&tab=overview) and [noaa-goes17](https://s3.console.aws.amazon.com/s3/buckets/noaa-goes17/?region=us-west-2&tab=overview) buckets in Amazon S3), where a "batch" is defined by user input, and persist to a local directory. Possible user input includes: satellites, channels, regions, or date range.

Should take special care thinking about performance, most often use-case, the local directory structure. 

Does not need to consider if a file has already been downloaded. 

## Developer Notes

`def download_batch(satellites, channels, regions, start_time, end_time, local_directory) => list[file_paths]`

```
def make_necessary_directories(filepath):
    """Create any directories in `filepath` that don't exist.
    Parameters
    ----------
    filepath : str
    """
    _logger.debug("Making necessary directories for %s", filepath)
    os.makedirs(name=os.path.dirname(filepath), exist_ok=True)


def download_scan(s3_bucket, s3_key, local_directory):
    """Download specific scan from S3.

    Parameters
    ----------
    s3_bucket : str
    s3_key : str
    local_directory : str

    Returns
    -------
    str
        File path scan was saved to.
    """
    s3 = boto3.client("s3")
    local_filepath = utilities.build_local_path(
        local_directory=local_directory, filepath=s3_key, satellite=s3_bucket
    )
    make_necessary_directories(filepath=local_filepath)
    _logger.info(
        "Downloading s3://%s/%s to %s", s3_bucket, s3_key, local_filepath,
    )
    s3.download_file(Bucket=s3_bucket, Key=s3_key, Filename=local_filepath)
    return local_filepath


def _is_good_object(key, regions, channels, start, end):
    region, channel, _, started_at = utilities.parse_filepath(filepath=key)
    return (region in regions) and (channel in channels) and (start <= started_at <= end)


def _num_hours_to_check(start, end):
    return math.ceil((end - start).total_seconds() / 3600)


def query(satellite, regions, channels, start, end):
    """Query Amazon S3 for data files that match the input.

    Parameters
    ----------
    satellite : str
    regions : list(str)
    channels : list(int)
    start : datetime.datetime
    end : datetime.datetime

    Returns
    -------
    list(boto3.resources.factory.s3.ObjectSummary)
    """
    _logger.info(
        """Querying for s3 objects with the following properties:
    satellite: %s
    regions: %s
    channels: %s
    start date: %s
    end date: %s""",
        satellite,
        regions,
        channels,
        start,
        end,
    )
    s3 = boto3.resource("s3")
    key_path_format = "{product_description}/{year}/{day_of_year}/{hour}/"
    product_description_format = "ABI-L1b-Rad{region}"

    scans = []
    for region in tqdm(regions, desc="Regions"):
        # only use the first character from region (M1 or M2 -> M)
        product_description = product_description_format.format(region=region[0])

        current = start.replace(minute=0, second=0, microsecond=0)
        inner_pbar = tqdm(total=_num_hours_to_check(start=current, end=end), desc="Hours")
        while current <= end:
            key_filter = key_path_format.format(
                product_description=product_description,
                year=current.year,
                day_of_year=current.strftime("%j"),
                hour=current.strftime("%H"),
            )
            s3_scans = [
                s3_object
                for s3_object in s3.Bucket(satellite).objects.filter(Prefix=key_filter)
                if _is_good_object(
                    key=s3_object.key,
                    regions=regions,
                    channels=channels,
                    start=start,
                    end=end,
                )
            ]
            scans += s3_scans
            current += datetime.timedelta(hours=1)
            inner_pbar.update(1)
        inner_pbar.close()
    return scans
```
## Sub-Tasks

- Team agreement on what user input to accept
- Team agreement on local directory structure
- Thoughtfully decide on how to filter objects in S3 for performance given most likely use-cases

## Acceptance Criteria

- [ ] Sub-tasks are completed
- [x] Successful batch download
- [x] Tests for non-boto3 functionality

### PR DESCRIPTION
**Pull request by camenpihor**
Fixes #10 

First pass for us at downloading a batch of scans from S3, I think we should go through this together. 

## Changes

- Downloads a batch of scans from S3, where batch is the combination of satellite, set of regions, set of channels, and date range.
- The local file structure of the download matches that of the file structure in S3, so that we can re-use code for navigating the S3 directory path and the local directory path.

### CODE REVIEW

File: tests/unit/test_downloader.py [hunk 1]
```diff
@@ -10,3 +12,50 @@ def test_make_necessary_directories():
...
+        start=datetime.datetime(2019, 12, 1, 1, 30),
+        end=datetime.datetime(2019, 12, 1, 10),
+    )
+    assert actual == 9
+
+    actual = downloader._num_hours_to_check(
+        start=datetime.datetime(2019, 12, 1, 1, 30),
+        end=datetime.datetime(2019, 12, 2, 10),
+    )
+    assert actual == 24 + 9
```
Comment (Ferrumofomega) [inline tests/unit/test_downloader.py]: Why "24 + 9" and not "33"?

File: wildfire/goes/downloader.py [hunk 2]
```diff
@@ -44,3 +47,150 @@ def download_scan(s3_bucket, s3_key, local_directory):
     )
     s3.download_file(Bucket=s3_bucket, Key=s3_key, Filename=local_filepath)
     return local_filepath
+
+
+def download_batch(satellite, regions, channels, start, end, local_directory):
```
Comment (Ferrumofomega) [inline wildfire/goes/downloader.py]: Why did you back off the ENUMs?

File: wildfire/goes/downloader.py [hunk 3]
```diff
@@ -44,3 +47,150 @@ def download_scan(s3_bucket, s3_key, local_directory):
...
+    key_path_format = "{product_description}/{year}/{day_of_year}/{hour}/"
+    product_description_format = "ABI-L1b-Rad{region}"
+
+    scans = []
+    for region in tqdm(regions, desc="Regions"):
+        # only use the first character from region (M1 or M2 -> M)
+        product_description = product_description_format.format(region=region[0])
+
+        current = start.replace(minute=0, second=0, microsecond=0)
+        inner_pbar = tqdm(total=_num_hours_to_check(start=current, end=end), desc="Hours")
```
Comment (Ferrumofomega) [inline wildfire/goes/downloader.py]: This variable name seems confusing to me, why "pbar"?

File: tests/unit/test_downloader.py (hunk 1 above)
Comment (camenpihor) [reply tests/unit/test_downloader.py]: just to kind of show that its 1 day + 9 hours, and not some magic number haha

File: wildfire/goes/downloader.py (hunk 3 above)
Comment (camenpihor) [reply wildfire/goes/downloader.py]: Yea! I can expand it to `progress_bar`

File: wildfire/goes/downloader.py (hunk 2 above)
Comment (camenpihor) [reply wildfire/goes/downloader.py]: Umm I would think about adding back, but I just decided that I would rather underbuild than overbuild right now

File: wildfire/goes/downloader.py (hunk 3 above)
Comment (camenpihor) [reply wildfire/goes/downloader.py]: done!

**Comment by camenpihor** [comment]
Going to merge for now and update the issue if there is any work left to do. 

This code is complete and tested

**Pull request merged**